app/src/main/kotlin/social/plasma/relay/RelayWebSocketListener.kt

```diff
@@ -0,0 +1,57 @@
+package social.plasma.relay
+
+import android.util.Log
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.runBlocking
+import okhttp3.Response
+import okhttp3.WebSocket
+import okhttp3.WebSocketListener
+import okio.ByteString
+import java.util.*
+import kotlin.coroutines.CoroutineContext
+
+class RelayWebSocketListener(
+    private val sink: MutableSharedFlow<String>,
+    private val coroutineContext: CoroutineContext
+) : WebSocketListener() {
+
+    private val subscriptionId = "plasma-sub-${UUID.randomUUID()}"
+
+    override fun onOpen(webSocket: WebSocket, response: Response) {
+        webSocket.send("""["REQ","$subscriptionId",{}]""")
+    }
+
+    /** Invoked when a text (type `0x1`) message has been received. */
+    override fun onMessage(webSocket: WebSocket, text: String) {
+        runBlocking(context = coroutineContext) { sink.emit(text) }
+    }
+
+    /** Invoked when a binary (type `0x2`) message has been received. */
+    override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
+        runBlocking(context = coroutineContext) { sink.emit(bytes.base64()) }
+    }
+
+    /**
+     * Invoked when the remote peer has indicated that no more incoming messages will be transmitted.
+     */
+    override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
+        Log.d(javaClass.simpleName, "closing: $reason")
```

Comment:
Not sure the android way to use logging. This seem ok?

Reply:
This is totally fine. We can add something more complex down the road